### PR TITLE
gh-73965: New environment variable PYTHON_HISTORY

### PR DESCRIPTION
**Pull request by ZackerySpytz**
It can be used to set the location of a .python_history file.

Co-Authored-By: Levi Sabah <0xl3vi@gmail.com>

* Issue: gh-73965

### CODE REVIEW

File: Doc/using/cmdline.rst [hunk 1]
```diff
@@ -918,6 +918,13 @@ conflict.
    .. versionadded:: 3.7
       See :pep:`540` for more details.
 
+.. envvar:: PYTHONHISTORY
+
+   This environment variable can be used to set the location of a
+   ``.python_history`` file (it is ``~/.python_history`` by default).
```
Comment (pablogsal) [inline Doc/using/cmdline.rst]: `~/.python_history` is for Unix but `%USERPROFILE%\.python_history` is for Windows (if memory does not fail me).

Also, there is a similar case in the docstring of `gethistoryfile()`

File: Lib/site.py [hunk 2]
```diff
@@ -393,6 +393,17 @@ def setcopyright():
 def sethelper():
     builtins.help = _sitebuiltins._Helper()
 
+def gethistoryfile():
+    """Check if the PYTHONHISTORY environment variable is set and define
+    it as the .python_history file.  If PYTHONHISTORY is not set, use the
+    default ~/.python_history file.
+    """
+    h = os.environ.get("PYTHONHISTORY")
```
Comment (pablogsal) [inline Lib/site.py]: Nit: can we use better variable names even if is obvious?

File: Lib/site.py [hunk 3]
```diff
@@ -393,6 +393,17 @@ def setcopyright():
...
+    if h:
```
Comment (pablogsal) [inline Lib/site.py]: Should we distinguish the case where is set but empty?  (`if h is not None`)

File: Misc/python.man [hunk 4]
```diff
@@ -502,6 +502,9 @@ show how long each import takes. This is exactly equivalent to setting
 .IP PYTHONBREAKPOINT
 If this environment variable is set to 0, it disables the default debugger. It
 can be set to the callable of your debugger of choice.
+.IP PYTHONHISTORY
+This environment variable can be used to set the location of a history file
+(it is \fI~/.python_history\fP by default).
```
Comment (pablogsal) [inline Misc/python.man]: Please, add the windows default as well.

**Comment by tirkarthi** [comment]
There are few other places in Docs where hardcoded value of `~/.python_history` is referred to.

```
$ rg python_history Doc
Doc/tutorial/interactive.rst
28:history into a file named :file:`.python_history` in your user directory.

Doc/library/readline.rst
292:file named :file:`.python_history` from the user's home directory.  The code
300:   histfile = os.path.join(os.path.expanduser("~"), ".python_history")
319:   histfile = os.path.join(os.path.expanduser("~"), ".python_history")

Doc/library/site.rst
135::file:`~/.python_history` as the history save file.  To disable it, delete (or
```

**Comment by ZackerySpytz** [comment]
@tirkarthi The first three cases mention `.python_history`, not `~/.python_history`, no?

The fourth case mentions `~/.python_history` in the context of the readline
module, which is not available on Windows. So, I'm not sure if those parts of
the docs need to be modified.

File: Misc/python.man (hunk 4 above)
Comment (ZackerySpytz) [reply Misc/python.man]: To be frank, I'm not sure if the Windows default should be mentioned in the man page...

File: Lib/site.py (hunk 3 above)
Comment (ZackerySpytz) [reply Lib/site.py]: I don't think so. Did you have something in mind?

**Comment by ZackerySpytz** [comment]
By the way, on Windows, `os.path.expanduser("~")` returns (for example) `'C:\\Users\\Tim'`.

**Review by tiran — DISMISSED** [review]
The new env var does not obey the ``sys.flags.ignore_environment`` flag. Please add a test that the new env var is ignored with ``-E`` or ``-I`` mode.

**Comment by tiran** [comment]
Could you please reword to explain what the history file actually does? It's non-obvious to users that are not familiar with readline module internals.

File: Lib/site.py [hunk 5]
```diff
@@ -428,13 +439,9 @@ def register_readline():
             pass
 
         if readline.get_current_history_length() == 0:
-            # If no history was loaded, default to .python_history.
```
Comment (tiran) [inline Lib/site.py:470]: please keep this original comment, too.

**Comment by ZackerySpytz** [comment]
I have made the requested changes; please review again.

**Comment by ZackerySpytz** [comment]
> Could you please reword to explain what the history file actually does?

Reword what?

**Comment by vszakats** [comment]
Can this be merged? It seems fairly obvious what `.python_history` means and it'd be quite nice to be able to control where it goes.

**Comment by EmilRehnberg** [comment]
Thanks for this!

Please make sure there will be no problem with paths with subdirectories that might not be created. E.g. `"export PYTHONHISTORY=$HOME/.cache/python/history` please assert that the path will be created if missing.

**Comment by wren** [comment]
Any way to help get this through? It looks like it's been ready for a while.

**Comment by ZLima12** [comment]
I'd like to see this merged as well; if there's anything I can do, I'd be happy to help.

**Comment by nash-pillai** [comment]
I would like this feature very much. @tiran and @pablogsal, please review this soon.

**Comment by eabase** [comment]
Guys, why on earth are you `hard-coding` a commonly used and often changed file-path like this?  

I was sincerely expecting to find an internal built-in variable named something like **`DEFAULT_HISTORY_FILE`**, but to my unpleasant surprise, I instead find **`history = os.path.join(os.path.expanduser('~'),'.python_history')`**.

File: Misc/python.man (hunk 4 above)
Comment (eabase) [reply Misc/python.man]: There is nothing wrong with mentioning Windows ops in `man` pages.

File: Doc/using/cmdline.rst [hunk 6]
```diff
@@ -918,6 +918,14 @@ conflict.
    .. versionadded:: 3.7
       See :pep:`540` for more details.
 
+.. envvar:: PYTHONHISTORY
```
Comment (hugovk) [inline Doc/using/cmdline.rst]: New env vars added in 3.13+ need an underscore: 

https://discuss.python.org/t/change-environment-variable-style/35180?u=hugovk

```suggestion
.. envvar:: PYTHON_HISTORY
```

File: Doc/using/cmdline.rst [hunk 7]
```diff
@@ -918,6 +918,14 @@ conflict.
...
+
+   This environment variable can be used to set the location of a
+   ``.python_history`` file (by default, it is ``.python_history`` in the
+   user's home directory).
+
+   .. versionadded:: 3.8
```
Comment (hugovk) [inline Doc/using/cmdline.rst]:
```suggestion
   .. versionadded:: 3.13
```

File: Doc/whatsnew/3.8.rst [hunk 8]
```diff
@@ -298,6 +298,9 @@ Other Language Changes
 * Added new ``replace()`` method to the code type (:class:`types.CodeType`).
   (Contributed by Victor Stinner in :issue:`37032`.)
 
+* The new :envvar:`PYTHONHISTORY` environment variable can be used to change
```
Comment (hugovk) [inline Doc/whatsnew/3.8.rst]:
```suggestion
* The new :envvar:`PYTHON_HISTORY` environment variable can be used to change
```
And move this to `Doc/whatsnew/3.13.rst`

File: Lib/site.py [hunk 9]
```diff
@@ -393,6 +393,18 @@ def setcopyright():
 def sethelper():
     builtins.help = _sitebuiltins._Helper()
 
+def gethistoryfile():
+    """Check if the PYTHONHISTORY environment variable is set and define
+    it as the .python_history file.  If PYTHONHISTORY is not set, use the
+    default .python_history file.
+    """
+    if not sys.flags.ignore_environment:
+        history = os.environ.get("PYTHONHISTORY")
```
Comment (hugovk) [inline Lib/site.py]:
```suggestion
    """Check if the PYTHON_HISTORY environment variable is set and define
    it as the .python_history file.  If PYTHON_HISTORY is not set, use the
    default .python_history file.
    """
    if not sys.flags.ignore_environment:
        history = os.environ.get("PYTHON_HISTORY")
```

File: Lib/site.py [hunk 10]
```diff
@@ -428,13 +440,13 @@ def register_readline():
             pass
 
         if readline.get_current_history_length() == 0:
-            # If no history was loaded, default to .python_history.
+            # If no history was loaded, default to .python_history,
+            # or PYTHONHISTORY.
```
Comment (hugovk) [inline Lib/site.py]:
```suggestion
            # or PYTHON_HISTORY.
```

File: Lib/test/test_site.py [hunk 11]
```diff
@@ -311,6 +312,19 @@ def test_no_home_directory(self):
...
+        filename = 'file'
+        rc, out, err = assert_python_ok('-c',
+            f'import site; assert site.gethistoryfile() == "{filename}"',
+            PYTHONHISTORY=filename)
+        self.assertEqual(rc, 0)
+
+        # Check that PYTHONHISTORY is ignored in isolated mode.
+        rc, out, err = assert_python_ok('-I', '-c',
+            f'import site; assert site.gethistoryfile() != "{filename}"',
+            PYTHONHISTORY=filename)
```
Comment (hugovk) [inline Lib/test/test_site.py]:
```suggestion
            PYTHON_HISTORY=filename)
        self.assertEqual(rc, 0)

        # Check that PYTHON_HISTORY is ignored in isolated mode.
        rc, out, err = assert_python_ok('-I', '-c',
            f'import site; assert site.gethistoryfile() != "{filename}"',
            PYTHON_HISTORY=filename)
```

File: Misc/NEWS.d/next/Library/2019-05-08-13-14-11.bpo-29779.jg33dp.rst [hunk 12]
```diff
@@ -0,0 +1,2 @@
+Add a new :envvar:`PYTHONHISTORY` environment variable to set the location
```
Comment (hugovk) [inline Misc/NEWS.d/next/Library/2019-05-08-13-14-11.bpo-29779.jg33dp.rst]:
```suggestion
Add a new :envvar:`PYTHON_HISTORY` environment variable to set the location
```

File: Misc/python.man [hunk 13]
```diff
@@ -502,6 +502,9 @@ show how long each import takes. This is exactly equivalent to setting
 .IP PYTHONBREAKPOINT
 If this environment variable is set to 0, it disables the default debugger. It
 can be set to the callable of your debugger of choice.
+.IP PYTHONHISTORY
```
Comment (hugovk) [inline Misc/python.man]:
```suggestion
.IP PYTHON_HISTORY
```

File: Python/initconfig.c [hunk 14]
```diff
@@ -95,7 +95,8 @@ static const char usage_6[] =
 "PYTHONBREAKPOINT: if this variable is set to 0, it disables the default\n"
 "   debugger. It can be set to the callable of your debugger of choice.\n"
 "PYTHONDEVMODE: enable the development mode.\n"
-"PYTHONPYCACHEPREFIX: root directory for bytecode cache (pyc) files.\n";
+"PYTHONPYCACHEPREFIX: root directory for bytecode cache (pyc) files.\n"
+"PYTHONHISTORY: the location of a .python_history file.\n";
```
Comment (hugovk) [inline Python/initconfig.c]:
```suggestion
"PYTHON_HISTORY: the location of a .python_history file.\n";
```

**Review by hugovk — COMMENTED** [review]
Please resolve the merge conflicts.

**Comment by hugovk** [comment]
I've resolved conflicts, renamed to `PYTHON_HISTORY` and targeted 3.13.

**Review by hauntsaninja — APPROVED** [review]

**Comment by hauntsaninja** [comment]
Thanks Hugo!

File: Python/initconfig.c [hunk 15]
```diff
@@ -296,6 +296,7 @@ static const char usage_envvars[] =
 "PYTHON_COLORS           : If this variable is set to 1, the interpreter will"
 "   colorize various kinds of output. Setting it to 0 deactivates this behavior.\n"
 "These variables have equivalent command-line parameters (see --help for details):\n"
+"PYTHON_HISTORY: the location of a .python_history file.\n"
```
Comment (serhiy-storchaka) [inline Python/initconfig.c:299]: What command line parameter sets it?

**Comment by llkj1** [comment]
感谢您与我们联系。(Thank you for contacting us.)
------------------------------------------------------------

我们收到了您的请求，并会在一至两个工作日内给您回复。您的案例编号是 102195504601。

想了解更多与开发相关主题的信息，请访问 开发者支持网页 (https://developer.apple.com/support/)。
                  
Apple Developer Program Support


We’ve received your support request and will get back to you in one to two business days. Your case number is 102195504601.

For additional information on development-related topics, visit:
https://developer.apple.com/support/

Best regards, 

Apple Developer Program Support
           

------------------------------------------------------------
Copyright (c) 2024 Apple Inc. All rights reserved.

Contact Us
https://developer.apple.com/contact/

Developer
https://developer.apple.com/

My Apple ID
https://appleid.apple.com

Privacy Policy
https://www.apple.com/privacy/

File: Python/initconfig.c (hunk 15 above)
Comment (hugovk) [reply Python/initconfig.c:299]: It doesn't have one, I compared with `PYTHON_COLORS` which doesn't either (https://github.com/python/cpython/pull/112732). Should we add one?

Comment (serhiy-storchaka) [reply Python/initconfig.c:299]: If it doesn't have one, then it is in the wrong section.

Comment (hugovk) [reply Python/initconfig.c:299]: Good point, please see PR https://github.com/python/cpython/pull/113798 to move it to the correct section.

**Pull request merged**